File: cloud-picture-backend/src/main/java/com/zack/cloudpicturebackend/controller/UserController.java
package com.zack.cloudpicturebackend.controller;


import cn.hutool.core.bean.BeanUtil;
import com.baomidou.mybatisplus.extension.plugins.pagination.Page;
import com.zack.cloudpicturebackend.annotation.AuthCheck;
import com.zack.cloudpicturebackend.common.CommonResult;
import com.zack.cloudpicturebackend.common.DeleteRequest;
import com.zack.cloudpicturebackend.constant.UserConstant;
import com.zack.cloudpicturebackend.exception.BusinessException;
import com.zack.cloudpicturebackend.exception.ErrorCode;
import com.zack.cloudpicturebackend.exception.ThrowUtil;
import com.zack.cloudpicturebackend.model.dto.user.*;
import com.zack.cloudpicturebackend.model.entity.User;
import com.zack.cloudpicturebackend.model.vo.LoginUserVO;
import com.zack.cloudpicturebackend.model.vo.UserVO;
import com.zack.cloudpicturebackend.service.UserService;
import org.springframework.web.bind.annotation.*;

import javax.annotation.Resource;
import javax.servlet.http.HttpServletRequest;
import java.util.List;

@RestController
@RequestMapping("/user")
public class UserController {
    @Resource
    UserService userService;

    @PostMapping("/register")
    public CommonResult<Long> registerUser(@RequestBody UserRegisterRequest userRegisterRequest){
        String userAccount = userRegisterRequest.getUserAccount();
        String password = userRegisterRequest.getPassword();
        String checkPassword = userRegisterRequest.getCheckPassword();
        long result = userService.userRegister(userAccount, password, checkPassword);
        return CommonResult.success(result);
    }

    @PostMapping("/login")
    public CommonResult<LoginUserVO> login(@RequestBody UserLoginRequest userLoginRequest, HttpServletRequest request){
        ThrowUtil.throwIf(userLoginRequest == null, ErrorCode.PARAMS_ERROR);
        String userAccount = userLoginRequest.getUserAccount();
        String password = userLoginRequest.getUserPassword();
        LoginUserVO result = userService.userLogin(userAccount, password,request);

        return CommonResult.success(result);
    }



    @GetMapping("/getLoginUser")
    public CommonResult<LoginUserVO> getLoginUser(HttpServletRequest request){
        User loginUser = userService.getLoginUser(request);
        String ds="dadad";
        return CommonResult.success(userService.getLoginUserVO(loginUser));
    }

    @GetMapping("/logout")
    public CommonResult<Boolean> logout(HttpServletRequest request){
        return CommonResult.success(userService.userLogout(request));
    }

    /**
     * 新增用户信息
     * @param userAddRequest
     * @return
     */
    @PostMapping("/addUser")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public CommonResult<Long> addUser(@RequestBody UserAddRequest userAddRequest){
        ThrowUtil.throwIf(userAddRequest == null, ErrorCode.PARAMS_ERROR);
        User user = new User();
        BeanUtil.copyProperties(userAddRequest, user);
        final String DEFAULT_PASSWORD = "123456";
        user.setUserPassword(userService.getEncryptPassword(DEFAULT_PASSWORD));
        boolean result = userService.save(user);
        ThrowUtil.throwIf(!result, ErrorCode.OPERATION_ERROR);
        return CommonResult.success(user.getId());
    }

    /**
     * 根据id获取用户信息  仅限管理员
     * @param id
     * @return
     */
    @GetMapping("/get")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public CommonResult<User> getUserById(long id){
        ThrowUtil.throwIf(id <= 0, ErrorCode.PARAMS_ERROR);
        User user = userService.getById(id);
        ThrowUtil.throwIf(user == null, ErrorCode.NOT_FOUND_ERROR);
        return CommonResult.success(user);
    }


    /**
     * 根据id获取封装用户
     * @param id
     * @return
     */
    @GetMapping("/get/vo")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public CommonResult<UserVO> getUserVOById(long id){
        ThrowUtil.throwIf(id <= 0, ErrorCode.PARAMS_ERROR);
        User data = this.getUserById(id).getData();
        UserVO userVO = userService.getUserVO(data);
        return CommonResult.success(userVO);
    }

    /**
     * 根据id删除用户
     * @param deleteRequest
     * @return
     */
    @DeleteMapping("/delete")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public CommonResult<Boolean> deleteUserById(@RequestBody DeleteRequest deleteRequest){
        if (deleteRequest == null||deleteRequest.getId()<=0){
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        boolean result=userService.removeById(deleteRequest.getId());
        return CommonResult.success(result);
    }

    /**
     * 更新用户信息
     * @param userUpdateRequest
     * @return
     */
    @PostMapping("/update")
    @AuthCheck(mustRole = UserConstant.ADMIN_ROLE)
    public CommonResult<Boolean> updateUser(@RequestBody UserUpdateRequest userUpdateRequest){
        if (userUpdateRequest == null||userUpdateRequest.getId()==null){
            throw new BusinessException(ErrorCode.PARAMS_ERROR);
        }
        User user = BeanUtil.copyProperties(userUpdateRequest, User.class);
        boolean result=userService.updateById(user);
        return CommonResult.success(result);
    }

    /**分页获取封装用户列表
     * @param userQueryRequest
     */
    @PostMapping("/list/page/vo")
    public CommonResult<Page<UserVO>> listUserVOByPage(@RequestBody UserQueryRequest userQueryRequest){
        ThrowUtil.throwIf(userQueryRequest == null, ErrorCode.PARAMS_ERROR);
        Integer current = userQueryRequest.getCurrent();
        Integer pageSize = userQueryRequest.getPageSize();

        Page<User> page = userService.page(new Page<>(current, pageSize), userService.getWrapper(userQueryRequest));

        List<UserVO> userVOList = userService.getUserVOList(page.getRecords());
        Page<UserVO> voPage = new Page<>(current,pageSize,page.getTotal());
        voPage.setRecords(userVOList);
        return CommonResult.success(voPage);
    }
}

File: cloud-picture-frontend/src/api/userController.ts
// @ts-ignore
/* eslint-disable */
import request from '@/request'

/** addUser POST /api/user/addUser */
export async function addUserUsingPost(body: API.UserAddRequest, options?: { [key: string]: any }) {
  return request<API.CommonResultLong_>('/api/user/addUser', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** deleteUserById DELETE /api/user/delete */
export async function deleteUserByIdUsingDelete(
  body: API.DeleteRequest,
  options?: { [key: string]: any }
) {
  return request<API.CommonResultBoolean_>('/api/user/delete', {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** getUserById GET /api/user/get */
export async function getUserByIdUsingGet(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getUserByIdUsingGETParams,
  options?: { [key: string]: any }
) {
  return request<API.CommonResultUser_>('/api/user/get', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  })
}

/** getUserVOById GET /api/user/get/vo */
export async function getUserVoByIdUsingGet(
  // 叠加生成的Param类型 (非body参数swagger默认没有生成对象)
  params: API.getUserVOByIdUsingGETParams,
  options?: { [key: string]: any }
) {
  return request<API.CommonResultUserVO_>('/api/user/get/vo', {
    method: 'GET',
    params: {
      ...params,
    },
    ...(options || {}),
  })
}

/** getLoginUser GET /api/user/getLoginUser */
export async function getLoginUserUsingGet(options?: { [key: string]: any }) {
  return request<API.CommonResultLoginUserVO_>('/api/user/getLoginUser', {
    method: 'GET',
    ...(options || {}),
  })
}

/** listUserVOByPage POST /api/user/list/page/vo */
export async function listUserVoByPageUsingPost(
  body: API.UserQueryRequest,
  options?: { [key: string]: any }
) {
  return request<API.CommonResultPageUserVO_>('/api/user/list/page/vo', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** login POST /api/user/login */
export async function loginUsingPost(body: API.UserLoginRequest, options?: { [key: string]: any }) {
  return request<API.CommonResultLoginUserVO_>('/api/user/login', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** logout GET /api/user/logout */
export async function logoutUsingGet(options?: { [key: string]: any }) {
  return request<API.CommonResultBoolean_>('/api/user/logout', {
    method: 'GET',
    ...(options || {}),
  })
}

/** registerUser POST /api/user/register */
export async function registerUserUsingPost(
  body: API.UserRegisterRequest,
  options?: { [key: string]: any }
) {
  return request<API.CommonResultLong_>('/api/user/register', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}

/** updateUser POST /api/user/update */
export async function updateUserUsingPost(
  body: API.UserUpdateRequest,
  options?: { [key: string]: any }
) {
  return request<API.CommonResultBoolean_>('/api/user/update', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    data: body,
    ...(options || {}),
  })
}
